Initial implementation `result_large_err`

lukaslueg opened the pull request:
This is a shot at #6560, #4652, and #3884. The lint checks for `Result` being returned from functions/methods where the `Err` variant is larger than a configurable threshold (the default of which is 128 bytes). There has been some discussion around this, which I'll try to quickly summarize:

* A large `Err`-variant may force an equally large `Result` if `Err` is actually bigger than `Ok`.
* There is a cost involved in large `Result`, as LLVM may choose to `memcpy` them around above a certain size.
* We usually expect the `Err` variant to be seldomly used, but pay the cost every time.
* `Result` returned from library code has a high chance of bubbling up the call stack, getting stuffed into `MyLibError { IoError(std::io::Error), ParseError(parselib::Error), ...}`, exacerbating the problem.

This PR deliberately does not take into account comparing the `Ok` to the `Err` variant (e.g. a ratio, or one being larger than the other). Rather we choose an absolute threshold for `Err`'s size, above which we warn. The reason for this is that `Err`s probably get `map_err`'ed further up the call stack, and we can't draw conclusions from the ratio at the point where the `Result` is returned. A relative threshold would also be less predictable, while not accounting for the cost of LLVM being forced to generate less efficient code if the `Err`-variant is _large_ in absolute terms.

We lint private functions as well as public functions, as the perf-cost applies to in-crate code as well.

In order to account for type-parameters, I conjured up `fn approx_ty_size`. The function relies on `LateContext::layout_of` to compute the actual size, and in case of failure (e.g. due to generics) tries to come up with an "at least size". In the latter case, the size of obviously wrong, but the inspected size certainly can't be smaller than that. Please give the approach a heavy dose of review, as I'm not actually familiar with the type-system at all (read: I have no idea what I'm doing).

The approach does, however flimsy it is, allow us to successfully lint situations like

```rust
pub union UnionError<T: Copy> {
    _maybe: T,
    _or_perhaps_even: (T, [u8; 512]),
}

// We know `UnionError<T>` will be at least 512 bytes, no matter what `T` is
pub fn param_large_union<T: Copy>() -> Result<(), UnionError<T>> {
    Ok(())
}
```

I've given some refactoring to `functions/result_unit_err.rs` to re-use some bits. This is also the groundwork for #6409

The default threshold is 128 because of https://github.com/rust-lang/rust-clippy/issues/4652#issue-505670554

`lintcheck` does not trigger this lint for a threshold of 128. It does warn for 64, though.

The suggestion currently is the following, which is just a placeholder for discussion to be had. I did have the computed size in a `span_label`. However, that might cause both ui-tests here and lints elsewhere to become flaky wrt to their output (as the size is platform dependent).

```
error: the `Err`-variant returned via this `Result` is very large
  --> $DIR/result_large_err.rs:36:34
   |
LL | pub fn param_large_error<R>() -> Result<(), (u128, R, FullyDefinedLargeError)> {
   |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ The `Err` variant is unusually large, at least 128 bytes
```

changelog: Add [`result_large_err`] lint

rust-highfive commented:
r? @Alexendoo

(rust-highfive has picked a reviewer for you, use r? to override)

lukaslueg commented:
I'll deal with CI after review. As far as suggestions go, maybe:

`error: the Err-variant returned from this public function is very large`
`^^^^^^^^^^^^^ The Err-variant is at least [actual size] bytes` (is this a problem for automation?)
`help: Try reducing the size of UnionError<T>, e.g. by boxing elements or replacing it with Box<UnionError<T>>`

Alexendoo commented on clippy_lints/src/functions/result.rs:
```diff
@@ -0,0 +1,152 @@
+fn approx_ty_size<'tcx>(cx: &LateContext<'tcx>, ty: ty::Ty<'tcx>) -> u64 {
+    use rustc_middle::ty::layout::LayoutOf;
+    match (cx.layout_of(ty).map(|layout| layout.size.bytes()), ty.kind()) {
+        (Ok(size), _) => size,
```
Could include arrays with something like

```suggestion
        (Ok(size), _) => size,
        (Err(_), ty::Array(t, n)) => n.try_eval_usize(cx.tcx, cx.param_env).map_or(0, |len| len * approx_ty_size(cx, t))
```

lukaslueg replied:
👍 I didn't think for `[(T, u128); 10]` et al

Alexendoo commented on clippy_lints/src/functions/result.rs:
```diff
@@ -0,0 +1,152 @@
+    item_span: Span,
+) -> Option<(&'tcx hir::Ty<'tcx>, ty::Ty<'tcx>)> {
+    if !in_external_macro(cx.sess(), item_span)
+    && let hir::FnRetTy::Return(hir_ty) = decl.output
+    && let ty = hir_ty_to_ty(cx.tcx, hir_ty)
+    && is_type_diagnostic_item(cx, ty, sym::Result)
+    && let ty::Adt(_, substs) = ty.kind() {
```
Some formatting as rustfmt doesn't do let chains yet:

```suggestion
        && let hir::FnRetTy::Return(hir_ty) = decl.output
        && let ty = hir_ty_to_ty(cx.tcx, hir_ty)
        && is_type_diagnostic_item(cx, ty, sym::Result)
        && let ty::Adt(_, substs) = ty.kind()
    {
```

Alexendoo commented on clippy_lints/src/functions/result.rs:
```diff
@@ -0,0 +1,152 @@
+
+pub(super) fn check_item<'tcx>(cx: &LateContext<'tcx>, item: &hir::Item<'tcx>, large_err_threshold: u64) {
+    if let hir::ItemKind::Fn(ref sig, _generics, _) = item.kind
+        && let Some((hir_ty, err_ty)) = result_err_ty(cx, sig.decl, item.span) {
```
```suggestion
        && let Some((hir_ty, err_ty)) = result_err_ty(cx, sig.decl, item.span)
    {
```

Alexendoo commented on clippy_lints/src/functions/result.rs:
```diff
@@ -0,0 +1,152 @@
+pub(super) fn check_impl_item<'tcx>(cx: &LateContext<'tcx>, item: &hir::ImplItem<'tcx>, large_err_threshold: u64) {
+    // Don't lint if method is a trait's implementation, we can't do anything about those
+    if let hir::ImplItemKind::Fn(ref sig, _) = item.kind
+        && let Some((hir_ty, err_ty)) = result_err_ty(cx, sig.decl, item.span)
+          && trait_ref_of_method(cx, item.def_id).is_none() {
+            if cx.access_levels.is_exported(item.def_id) {
+                let fn_header_span = item.span.with_hi(sig.decl.output.span().hi());
+                check_result_unit_err(cx, err_ty, fn_header_span);
+            }
+            check_result_large_err(cx, err_ty, hir_ty.span, large_err_threshold);
+        }
```
```suggestion
        && let Some((hir_ty, err_ty)) = result_err_ty(cx, sig.decl, item.span)
        && trait_ref_of_method(cx, item.def_id).is_none()
    {
        if cx.access_levels.is_exported(item.def_id) {
            let fn_header_span = item.span.with_hi(sig.decl.output.span().hi());
            check_result_unit_err(cx, err_ty, fn_header_span);
        }
        check_result_large_err(cx, err_ty, hir_ty.span, large_err_threshold);
    }
```

Alexendoo reviewed (commented):
Looks good so far!

Might be worth grepping for `layout_of` and checking if any other usages of it could benefit from this `approx_ty_size` fn

lukaslueg commented:
All addressed. Will squash if so requested.

I've moved `approx_ty_size` to `clippy_utils::ty`, where it might be useful for other things like [`large_enum_variant`](https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant). I may grep and address that in a later PR. Added the `array`-case including `substs` and added tests for that. Modified the suggestion as indicated above, happy for comments.

Alexendoo commented on clippy_lints/src/functions/mod.rs:
````diff
@@ -217,17 +217,62 @@ declare_clippy_lint! {
+    /// ### Examples
+    /// ```rust
+    /// pub enum ParseError {
+    ///     DoesntMakeMuchSenseAnyway,
+    ///     UnparsedBytes([u8; 512]),
+    /// }
+    ///
+    /// // The `Result` has at least 512 bytes, even in the `Ok`-case
+    /// pub fn parse() -> Result<(), ParseError> {
+    ///     todo!()
+    /// }
+    /// ```
+    /// should be
+    /// ```
+    /// pub enum ParseError {
+    ///     DoesntMakeMuchSenseAnyway,
+    ///     UnparsedBytes(Box<[u8; 512]>),
+    /// }
+    ///
+    /// // The `Result` is slightly large than a pointer
+    /// pub fn parse() -> Result<(), ParseError> {
+    ///     todo!()
+    /// }
````
Couple of doc nits:

```suggestion
    ///     UnparsedBytes([u8; 512]),
    ///     UnexpectedEof,
    /// }
    ///
    /// // The `Result` has at least 512 bytes, even in the `Ok`-case
    /// pub fn parse() -> Result<(), ParseError> {
    ///     Ok(())
    /// }
    /// ```
    /// should be
    /// ```
    /// pub enum ParseError {
    ///     UnparsedBytes(Box<[u8; 512]>),
    ///     UnexpectedEof,
    /// }
    ///
    /// // The `Result` is slightly larger than a pointer
    /// pub fn parse() -> Result<(), ParseError> {
    ///     Ok(())
    /// }
```

Alexendoo reviewed (commented):
Sorry for the delay! Just a couple nits then if you squash the commits this should be ready to merge

lukaslueg commented:
Addressed the nit, squashed. One may decide to close #6560, #4652, and/or #3884 once merged.

lukaslueg commented:
*yells at CI for catching things `cargo test` doesn't*

Alexendoo commented:
Great, thanks for the lint!

I stuck a `changelog: ..` entry in the PR message

@bors r+

bors commented:
:pushpin: Commit 66a97055b2721700b472e4460a560c8ded70b3c2 has been approved by `Alexendoo`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=Alexendoo 66a97055b2721700b472e4460a560c8ded70b3c2 -->
<!-- homu: {"type":"Approved","sha":"66a97055b2721700b472e4460a560c8ded70b3c2","approver":"Alexendoo","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit 66a97055b2721700b472e4460a560c8ded70b3c2 with merge 09e4659a862f8a244fedf473a06b9a85448d3a17...
<!-- homu: {"type":"BuildStarted","head_sha":"66a97055b2721700b472e4460a560c8ded70b3c2","merge_sha":"09e4659a862f8a244fedf473a06b9a85448d3a17"} -->

lukaslueg commented:
@Alexendoo Thanks for the review

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/8098974073?check_suite_focus=true), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/8098946279?check_suite_focus=true), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/8099156352?check_suite_focus=true)
Approved by: Alexendoo
Pushing 09e4659a862f8a244fedf473a06b9a85448d3a17 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Alexendoo","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/8099156352?check_suite_focus=true","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/8098974073?check_suite_focus=true","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/8098946279?check_suite_focus=true"},"merge_sha":"09e4659a862f8a244fedf473a06b9a85448d3a17"} -->
